src/main/java/org/apache/sysds/runtime/instructions/fed/DataGenFEDInstruction.java

```diff
@@ -0,0 +1,366 @@
+			method = OpOpDG.RAND;
+			InstructionUtils.checkNumFields(s, 10, 11);
+		}
+		else
+			throw new DMLRuntimeException("DataGenFEDInstruction: only matrix rand(..) is supported.");
+
+		CPOperand out = new CPOperand(s[s.length - 1]);
+		Operator op = null;
+
+		if(method == OpOpDG.RAND) {
+			int missing; // number of missing params (row & cols or dims)
+			CPOperand rows = null, cols = null, dims = null;
+			if(s.length == 12) {
+				missing = 1;
+				rows = new CPOperand(s[1]);
+				cols = new CPOperand(s[2]);
+			}
+			else {
+				missing = 2;
+				dims = new CPOperand(s[1]);
+			}
+			int blen = Integer.parseInt(s[4 - missing]);
+			double sparsity = !s[7 - missing].contains(Lop.VARIABLE_NAME_PLACEHOLDER) ? Double
+				.parseDouble(s[7 - missing]) : -1;
+			long seed = !s[SEED_POSITION_RAND - missing].contains(Lop.VARIABLE_NAME_PLACEHOLDER) ? Long
+				.parseLong(s[SEED_POSITION_RAND - missing]) : -1;
+			String pdf = s[9 - missing];
+			String pdfParams = !s[10 - missing].contains(Lop.VARIABLE_NAME_PLACEHOLDER) ? s[10 - missing] : null;
+			int k = Integer.parseInt(s[11 - missing]);
+
+			return new DataGenFEDInstruction(op, method, null, out, rows, cols, dims, blen, s[5 - missing],
+				s[6 - missing], sparsity, seed, pdf, pdfParams, k, opcode, str);
+		}
+		else
+			throw new DMLRuntimeException("Unrecognized data generation method: " + method);
+	}
+
+	@Override
+	public void processInstruction(ExecutionContext ec) {
+		// process specific datagen operator
+		if(method == OpOpDG.RAND)
+			processRandInstruction(ec);
+	}
+
+	private MatrixObject processRandInstruction(ExecutionContext ec) {
+		MatrixObject out;
+
+		// derive number of workers to use
+		int nworkers = DMLScript.FED_WORKER_PORTS.size();
+		if(getRows() % nworkers != 0) {
+			if(Math.round(getRows() / (double) nworkers) == Math.floor(getRows() / (double) nworkers))
+				nworkers--;
+		}
+
+		// generate seeds and rows
+		long[] lSeeds = randomSeedsGenerator(nworkers);
+
+		// new sizes
+		int size = (int) Math.round(getRows() / (double) nworkers);
+		int addedRows = size * (nworkers -1);
+		long[] rowsPerWorker = new long[nworkers];
+		Arrays.fill(rowsPerWorker, 0, nworkers-1, size);
+		rowsPerWorker[nworkers-1] = getRows() - addedRows;
+
+		out = processRandInstructionMatrix(ec, rowsPerWorker, lSeeds);
+
+		// reset runtime seed (e.g., when executed in loop)
+		runtimeSeed = null;
+		return out;
+	}
+
+	private MatrixObject processRandInstructionMatrix(ExecutionContext ec, long[] rowsPerWorker, long[] lSeeds) {
+		long lrows = ec.getScalarInput(rows).getLongValue();
+		long lcols = ec.getScalarInput(cols).getLongValue();
+		checkValidDimensions(lrows, lcols);
+
+		String[] instStrings = new String[rowsPerWorker.length];
+		Arrays.fill(instStrings, 0, rowsPerWorker.length, instString);
+
+		InetAddress addr = null;
+		try {
+			addr = InetAddress.getLocalHost();
+		}
+		catch(UnknownHostException e) {
+			e.printStackTrace();
+		}
+		String host = addr.getHostName();
```

Comment:
With only here is another argument for why we need to specify a list of federated addresses, in the arguments for the instruction since we need to specify other hosts than localhost.